peps/pep-0747.rst

```diff
@@ -0,0 +1,977 @@
+         typedload.\ `load`_
+
+   -  Pattern 2:
+
+      ::
+
+        class Converter[T]:
+            def __init__(self, typx: TypeExpr[T]) -> None: ...
+            def convert(self, value: object) -> T: ...
+
+      -  Examples: pydantic.\ `TypeAdapter(T).validate_python`_,
+         mashumaro.\ `JSONDecoder(T).decode`_
+
+.. _structure: https://github.com/python-attrs/cattrs/blob/5f5c11627a7f67a23d6212bc7df9f96243c62dc5/src/cattrs/converters.py#L332-L334
+.. _checkcast: https://github.com/davidfstr/trycast#checkcast-api
+.. _load: https://ltworf.github.io/typedload/
+.. _TypeAdapter(T).validate_python: https://stackoverflow.com/a/61021183/604063
+.. _JSONDecoder(T).decode: https://github.com/Fatal1ty/mashumaro?tab=readme-ov-file#usage-example
+
+-  Typed field definitions:
+
+   -  Pattern:
+
+      ::
+
+        class Field[T]:
+            value_type: TypeExpr[T]
+
+   -  Examples: attrs.\ `make_class`_,
+      dataclasses.\ `make_dataclass`_ [#DataclassInitVar]_, `openapify`_
+
+.. _make_class: https://www.attrs.org/en/stable/api.html#attrs.make_class
+.. _make_dataclass: https://github.com/python/typeshed/issues/11653
+.. _openapify: https://github.com/Fatal1ty/openapify/blob/c8d968c7c9c8fd7d4888bd2ddbe18ffd1469f3ca/openapify/core/models.py#L16
+
+The survey also identified some introspection functions that take
+annotation expressions as input using plain ``object``\ s which would
+*not* gain functionality by marking those inputs as ``TypeExpr[]``:
+
+-  General introspection operations:
+
+   -  Pattern: ``def get_annotation_info(maybe_annx: object) -> object``
+   -  Examples: typing.{`get_origin`_, `get_args`_},
+      `typing_inspect`_.{is_*_type, get_origin, get_parameters}
+
+.. _get_origin: https://docs.python.org/3/library/typing.html#typing.get_origin
+.. _get_args: https://docs.python.org/3/library/typing.html#typing.get_args
+.. _typing_inspect: https://github.com/ilevkivskyi/typing_inspect?tab=readme-ov-file#readme
+
+
+Rationale
+=========
+
+Before this PEP existed there were already a few definitions in use to describe
+different kinds of type annotations:
+
+::
+
+   +----------------------------------+ 
+   | +------------------------------+ | 
+   | | +-------------------------+  | | 
+   | | | +---------------------+ |  | | 
+   | | | | Class object        | |  | | = type[C]        
+   | | | +---------------------+ |  | | 
+   | | | Type expression object  |  | | = TypeExpr[T]  <-- new!
+   | | +-------------------------+  | | 
+   | | Annotation expression object | | 
+   | +------------------------------+ | 
+   | Object                           | = object         
+   +----------------------------------+ 
+
+-  :ref:`Class objects <typing:type-brackets>`,
+   spelled as ``type[C]``, support ``isinstance`` checks and are callable.
+
+   -  Examples: ``int``, ``str``, ``MyClass``
+
+-  :ref:`Type expressions <typing:type-expression>`
+   include any type annotation which describes a type.
```

Comment:
```suggestion
   include any expression which describes a type.
```

Reply:
Both the "Type expressions" and "Annotation expressions" sections start with "X include any type annotation", so I'd like to keep them consistent.

I think it would be misleading to make the paired change to the "Annotation expressions" section, which would then read "Annotation expressions include any **expression**, including [...]" and be too general a statement.

Reply:
However, type expressions are not necessarily annotations; `T` in `cast(T, expr)` is a type expression.